File: turnitover/render/protocol.py
"""Python mirror of web/src/protocol.ts plus decoding helpers."""
from __future__ import annotations

import base64
from typing import Literal, TypedDict

import numpy as np

Vec3 = tuple[float, float, float]


class ViewDefDict(TypedDict, total=False):
    id: str
    azimuth_deg: float
    elevation_deg: float
    projection: Literal["perspective", "orthographic"]
    distance_factor: float


class FramingDict(TypedDict):
    center: list[float]
    radius: float


class LoadResponse(TypedDict):
    ok: bool
    parts: list[str]
    joints: list[dict]
    framing: FramingDict


class ImageResponse(TypedDict):
    image_png_b64: str
    width: int
    height: int
    render_ms: float


class StatsPayload(TypedDict):
    property: str
    triangles_rendered: int
    draw_calls: int
    triangles_scene: int
    mesh_count: int
    geometries: int
    textures: int
    programs: int
    last_render_ms: float


class PartGeometryDict(TypedDict):
    name: str
    vertex_count: int
    triangle_count: int
    positions_b64: str
    indices_b64: str


class GeometryExportDict(TypedDict):
    joint_state: dict[str, float]
    parts: list[PartGeometryDict]


STATS_KEYS = tuple(StatsPayload.__annotations__)


def decode_png(b64: str) -> bytes:
    return base64.b64decode(b64)


def decode_positions(b64: str) -> np.ndarray:
    return np.frombuffer(base64.b64decode(b64), dtype="<f4").reshape(-1, 3)


def decode_indices(b64: str) -> np.ndarray:
    return np.frombuffer(base64.b64decode(b64), dtype="<u4").reshape(-1, 3)

File: web/src/protocol.ts
// Request/response types of window.harness. Mirrored by turnitover/render/protocol.py.
export type Vec3 = [number, number, number];
export type Projection = "perspective" | "orthographic";

export interface Framing { center: Vec3; radius: number }

export interface ViewDef { id: string; azimuth_deg: number; elevation_deg: number; projection: Projection; distance_factor?: number }

export interface LoadRequest { source_js: string; framing: Framing | "auto"; width: number; height: number; views: ViewDef[] }
export interface LoadResponse {
  ok: true;
  parts: string[];
  joints: { id: string; type: string; part: string; limits: [number, number] }[];
  framing: Framing;
}

export interface CameraInfo { position: Vec3; target: Vec3; projection: Projection }
export interface ImageResponse { image_png_b64: string; width: number; height: number; render_ms: number }
export interface ViewResponse extends ImageResponse { view_id: string; camera: CameraInfo }
export interface ActuateResponse extends ImageResponse { joint_id: string; value: number }

export type RuntimeProperty = "stats" | "hierarchy" | "joint_state" | "state_delta";
export interface StatsPayload {
  property: "stats";
  triangles_rendered: number; draw_calls: number; triangles_scene: number; mesh_count: number;
  geometries: number; textures: number; programs: number; last_render_ms: number;
}
export interface HierarchyNode { name: string; type: string; position: Vec3; rotation: Vec3; scale: Vec3; children: HierarchyNode[] }
export interface HierarchyPayload { property: "hierarchy"; root: HierarchyNode }
export interface JointStatePayload { property: "joint_state"; joints: Record<string, number> }
export interface StateDeltaPayload {
  property: "state_delta"; since: string;
  parts: Record<string, { aabb_min_delta: Vec3; aabb_max_delta: Vec3 }>;
}
export type RuntimePayload = StatsPayload | HierarchyPayload | JointStatePayload | StateDeltaPayload;

export interface PartGeometry { name: string; vertex_count: number; triangle_count: number; positions_b64: string; indices_b64: string }
export interface GeometryExport { joint_state: Record<string, number>; parts: PartGeometry[] }

export interface HarnessError { stage: "import" | "create" | "validate" | "runtime"; message: string }
